Named graphs implementation

ArkanEmre opened the pull request:
-Import of RDF data sets from TriG and N-Quads files
-Export of RDF data sets into TriG and N-Quads files
-Quadruple-wise deletion of imported RDF data sets

Commit:
add RDFDatasetToLPGStatementProcessor, RDFDatasetDirectStatementLoader for import of RDF data sets
add importRDFDataset to ImportRDF
add new test and test file for RDF data set import

Commit:
Merge branch 'develop' of https://github.com/eccenca/neosemantics into NamedGraphs

jbarrasa committed:
Merge pull request #93 from jbarrasa/3.5

sync gh-pages with 3.5

Commit:
upgrade rdf4j version to 2.5.2 (latest)
add rdf4j-rio-nquads to dependencies
fix a test

Commit:
transform ContextedResource to regular class
rename ContexedResource to ContextResource
fix issue with retrieval of existing nodes during RDF data set import
add new tests and test files
add N-Quads to available parsers

Commit:
clean RDFDatasetToLPGStatementProcessor from redundant code

Commit:
rework class to avoid redundant code

Commit:
add support for export of RDF data sets to RDFEndpoint
add tests for export of RDF data sets

Commit:
add deleteRDFDataset method to RDFImport
add RDFDatasetDirectStatementDeleter for RDF data set deletion
add tests and test files for RDF data set deletion
fix bug in DirectStatementDeleter

Commit:
Merge branch 'develop' into NamedGraphs

# Conflicts:
#	src/main/java/semantics/extension/RDFEndpoint.java
#	src/test/java/semantics/RDFImportTest.java
#	src/test/java/semantics/extension/RDFEndpointTest.java

Commit:
fix ParseRDF call in importRDFDataset with new parameters
fix tests for RDFEndpoint with PathParams and QueryParams
fix bug in RDFEndpoint

Commit:
add trig files for import and deletion of RDF datasets (named graphs)
add documentation for import and deletion of RDF datasets (named graphs)

rnavarropiris commented on src/main/java/semantics/ContextResource.java:
```diff
@@ -0,0 +1,50 @@
+/**
+ * ContextResource contains a uri and a graphUri
+ * It represents a Resource with an optional graph (context) uri
+ * It is used as Key for the Maps containing labels and properties
```
Add a `@see` javadoc annotation pointing to the class using this one instead

rnavarropiris commented on src/main/java/semantics/ContextResource.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Objects;
+
+/**
+ * ContextResource contains a uri and a graphUri
```
I would delete this sentence and add javadoc for all getters of the class.

rnavarropiris commented on src/main/java/semantics/ContextResource.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class ContextResource {
+
+  private String uri;
```
This parameter is not intended to be modified, so it should be declared `final`

rnavarropiris commented on src/main/java/semantics/ContextResource.java:
```diff
@@ -0,0 +1,50 @@
+public class ContextResource {
+
+  private String uri;
+  private String graphUri;
```
This parameter is not intended to be modified, so it should be declared `final`

rnavarropiris commented on src/main/java/semantics/DirectStatementDeleter.java:
```diff
@@ -263,7 +258,17 @@ private void setbNodeInfo(String bNodeInfo) {
   }
 
-  private void addNamespaceNode() {
+  private void deleteNodeIfEmpty(Node node) {
```
Proposal for better readability:

```java
private void deleteNodeIfEmpty(Node node) {
    
    int nodePropertyCount = node.getAllProperties().size();
    int labelCount = Iterators.size(node.getLabels().iterator());
    
    if (!node.hasRelationship(Direction.OUTGOING) &&
        !node.hasRelationship(Direction.INCOMING) &&
        node.hasLabel(RESOURCE) && labelCount == 1 &&
        node.getAllProperties().containsKey("uri") && nodePropertyCount == 1) {
      node.delete();
    }
  }
```

rnavarropiris commented on src/main/java/semantics/RDFDatasetDirectStatementDeleter.java:
```diff
@@ -0,0 +1,349 @@
+    nodeCache = CacheBuilder.newBuilder()
+        .maximumSize(conf.getNodeCacheSize())
+        .build();
+    log = l;
```
This should be part of the super-constructor. Relying on the sub-classes to initialize this value is bad design.

Should be refactored, so that the constructor of `RDFDatasetToLPGStatementProcessor` requires a `Log` and sets its value on construction (ideally as `final`)

rnavarropiris commented on src/main/java/semantics/RDFDatasetDirectStatementDeleter.java:
```diff
@@ -0,0 +1,349 @@
+            return node;
+          }
+        });
+      } catch (InvalidCacheLoadException | IllegalStateException e) {
```
The method already declares `throws Exception`, so why is this catched?
Also, just printing an error will not stop the rest of the process, which will likely lead to further exceptions.

Therefore the try-catch wrapping should be removed.

ArkanEmre replied:
If the try-catch block is removed the whole deletion process will be interrupted even if only one node cannot be found. To be able to support repetitive deletion, this and the following try-catch blocks have to handle the exceptions thrown.

rnavarropiris replied:
Ok, fair enough.
Maybe we can discuss this tomorrow with Jesus, I think a more elegant solution than `e.printStackStrace()` is required at this point (e.g. a deletion report), but this is probably out-of-scope in this feature branch.

rnavarropiris commented on src/main/java/semantics/RDFDatasetDirectStatementDeleter.java:
```diff
@@ -0,0 +1,349 @@
+            return node;
+          }
+        });
+      } catch (InvalidCacheLoadException | NoSuchElementException e) {
```
Same here (regarding try-catch usage)

rnavarropiris commented on src/main/java/semantics/RDFDatasetDirectStatementDeleter.java:
```diff
@@ -0,0 +1,349 @@
+            return node;
+          }
+        });
+      } catch (InvalidCacheLoadException | NoSuchElementException e) {
```
Same here (regarding try-catch usage)

rnavarropiris commented on src/main/java/semantics/RDFDatasetDirectStatementDeleter.java:
```diff
@@ -0,0 +1,349 @@
+    mappedTripleCounter = 0;
+  }
+
+  Map<String, String> getNamespaces() {
```
Probably a case of bad design: this parameter belongs to the super-class, so the super-class should provide a getter for it.

rnavarropiris commented on src/main/java/semantics/RDFDatasetDirectStatementDeleter.java:
```diff
@@ -0,0 +1,349 @@
+    this.bNodeInfo = bNodeInfo;
+  }
+
+  private void deleteNodeIfEmpty(Node node) {
```
Proposal for better readability:

```java
private void deleteNodeIfEmpty(Node node) {

    int nodePropertyCount = node.getAllProperties().size();
    int labelCount = Iterators.size(node.getLabels().iterator());

    if (!node.hasRelationship(Direction.OUTGOING) &&
        !node.hasRelationship(Direction.INCOMING) &&
        node.hasLabel(RESOURCE) && labelCount == 1 &&
        node.getAllProperties().containsKey("uri") &&
        ((node.getAllProperties().containsKey("graphUri") && nodePropertyCount == 2) ||
            nodePropertyCount == 1)) {
      node.delete();
    }
  }
```

rnavarropiris commented on src/test/java/semantics/extension/RDFEndpointTest.java:
```diff
@@ -1528,6 +1525,322 @@ public Void apply(GraphDatabaseService graphDatabaseService) throws RuntimeExcep
+
+              tx.success();
+            } catch (Exception e) {
+              fail(e.getMessage());
```
The test fails if a non-expected exception is thrown, so this try-catch block is not only not necessary, but also (imo) counter-productive, as the stack-trace will not be shown in the error logs (leading to more difficult debugging if something goes wrong).

I would just let the exception be thrown (so just remove the catch block, the try-with-resources construction is still good)

rnavarropiris replied:
Alternatively, in case this is problematic (since the signature defines `RuntimeException` and not `Exception`, which would mean having to wrap the exceptions) I would at least execute `e.printStackTrace()` before the `fail` command to avoid information loss.

The same should apply to all other similar cases in the test class.

rnavarropiris commented on src/test/java/semantics/extension/RDFEndpointTest.java:
```diff
@@ -1528,6 +1525,322 @@ public Void apply(GraphDatabaseService graphDatabaseService) throws RuntimeExcep
+                  + "','TriG',{ handleVocabUris: 'KEEP', typesToLabels: true, commitSize: 500, keepCustomDataTypes: true, handleMultival: 'ARRAY'})");
+              tx.success();
+            } catch (Exception e) {
+              fail(e.getMessage());
```
Same case as in a few lines up.

rnavarropiris commented on src/test/java/semantics/extension/RDFEndpointTest.java:
```diff
@@ -1528,6 +1525,322 @@ public Void apply(GraphDatabaseService graphDatabaseService) throws RuntimeExcep
+
+              tx.success();
+            } catch (Exception e) {
+              fail(e.getMessage());
```
Same as in the previous test case.

rnavarropiris commented on src/test/java/semantics/extension/RDFEndpointTest.java:
```diff
@@ -1528,6 +1525,322 @@ public Void apply(GraphDatabaseService graphDatabaseService) throws RuntimeExcep
+                  + "','N-Quads',{ handleVocabUris: 'KEEP', typesToLabels: true, commitSize: 500, keepCustomDataTypes: true, handleMultival: 'ARRAY'})");
+              tx.success();
+            } catch (Exception e) {
+              fail(e.getMessage());
```
Same as in the previous test case.

rnavarropiris commented on src/test/java/semantics/extension/RDFEndpointTest.java:
```diff
@@ -1528,6 +1525,322 @@ public Void apply(GraphDatabaseService graphDatabaseService) throws RuntimeExcep
+
+              tx.success();
+            } catch (Exception e) {
+              fail(e.getMessage());
```
Same as in the previous test case.

rnavarropiris commented on src/test/java/semantics/extension/RDFEndpointTest.java:
```diff
@@ -1528,6 +1525,322 @@ public Void apply(GraphDatabaseService graphDatabaseService) throws RuntimeExcep
+
+              tx.success();
+            } catch (Exception e) {
+              fail(e.getMessage());
```
Same as in the previous test case.

rnavarropiris replied:
There are several more, I would apply the same change to all.

rnavarropiris commented on src/main/java/semantics/RDFDatasetDirectStatementDeleter.java:
```diff
@@ -0,0 +1,349 @@
+    for (Map.Entry<ContextResource, Set<String>> entry : resourceLabels.entrySet()) {
+      if (entry.getKey().getUri().startsWith("genid")) {
+        bNodeCount++;
+        notDeletedStatementCount++;
```
If the key of the node is a blank node, the whole node is skipped.

This means, the increase in `bNodeCount` makes sense, but what about `notDeletedStatementCount`? 
This seems inconsistent.

rnavarropiris commented on src/main/java/semantics/RDFDatasetDirectStatementDeleter.java:
```diff
@@ -0,0 +1,349 @@
+        notDeletedStatementCount++;
+        continue;
+      }
+      Node tempNode = null;
```
If the try-catch blocks are simplified, usage of this variable is unnecessary (just initialize `node` once)

rnavarropiris commented on src/main/java/semantics/RDFDatasetDirectStatementDeleter.java:
```diff
@@ -0,0 +1,349 @@
+        notDeletedStatementCount++;
+      }
+      entry.getValue().forEach(l -> {
+        if (node != null && node.hasLabel(Label.label(l))) {
```
I think the `node.hasLabel` check is not necessary: to my understanding `removeLabel` only removes if the label exists, does not fail otherwise.

rnavarropiris replied:
Ignore, just realized this is required for the consistency of the counter.

Commit:
edit javadoc comments in ContextResource
add final to uri and graphUri declarations

Commit:
initialize Log in RDFToLPGStatementProcessor instead of in the Subclasses
fix deleteNodeIfEmpty in RDFDatasetDirectStatementDeleter and DirectStatementDeleter for better readability

Commit:
fix deleteNodeIfEmpty in RDFDatasetDirectStatementDeleter for better readability

Commit:
move getNamespaces to RDFToLPGStatementProcessor from the Subclasses for consistency

Commit:
rename bNodeCount to statementsWithbNodeCount in DirectStatementDeleter and RDFDatasetDirectStatementDeleter
fix bug in DirectStatementDeleter and RDFDatasetDirectStatementDeleter where statementsWithbNodeCount delivered wrong values
add new assertions to some tests in RDFEndpointTest

Commit:
add e.printStackTrace() to unit tests in RDFEndpointTest to avoid information loss

Commit:
add configSummary to importRDFDataset

Commit:
add new test for repetitive deletion of RDF datasets
fix bug where notDeletedStatementCount delivered wrong results
change code structure for better readability

Commit:
Merge branch 'gh-pages' into NamedGraphs

Commit:
fix ContextResource javadoc, minor syntax improvements
